libs/libc/machine/arm64/arch_elf.c

```diff
@@ -0,0 +1,789 @@
+                               INSN_IMM_ADR);
+        }
+        break;
+
+      case R_AARCH64_ADR_PREL_PG_HI21_NC:
+        {
+          overflow_check = false;
+        }
+
+      case R_AARCH64_ADR_PREL_PG_HI21:
+        {
+          if (((uint64_t)addr & 0xfff) < 0xff8)
+            {
+              ret = reloc_insn_imm(RELOC_OP_PAGE, addr, val, 12, 21,
+                                   INSN_IMM_ADR);
+            }
+          else
+            {
+              uint32_t insn;
+
+              /* patch ADRP to ADR if it is in range */
+
+              ret = reloc_insn_imm(RELOC_OP_PREL, addr, val & ~0xfff, 0, 21,
+                                   INSN_IMM_ADR);
+              if (ret == 0)
+                {
+                  insn = le32toh(*(uint32_t *)addr);
+                  insn &= ~BIT(31);
+                  *(uint32_t *)addr = htole32(insn);
+                }
+              else
+                {
+                  berr("Out of range for ADR\n");
+                  return -EINVAL;
+                }
+            }
+        }
+        break;
+
+      case R_AARCH64_ADD_ABS_LO12_NC:
+      case R_AARCH64_LDST8_ABS_LO12_NC:
+        {
+          overflow_check = false;
+          ret = reloc_insn_imm(RELOC_OP_ABS, addr, val, 0, 12,
+                               INSN_IMM_12);
+        }
+        break;
+
+      case R_AARCH64_LDST16_ABS_LO12_NC:
+        {
+          overflow_check = false;
+          ret = reloc_insn_imm(RELOC_OP_ABS, addr, val, 1, 11,
+                               INSN_IMM_12);
+        }
+        break;
+
+      case R_AARCH64_LDST32_ABS_LO12_NC:
+        {
+          overflow_check = false;
+          ret = reloc_insn_imm(RELOC_OP_ABS, addr, val, 2, 10,
+                               INSN_IMM_12);
+        }
+        break;
+
+      case R_AARCH64_LDST64_ABS_LO12_NC:
+        {
+          overflow_check = false;
+          ret = reloc_insn_imm(RELOC_OP_ABS, addr, val, 3, 9,
+                               INSN_IMM_12);
+        }
+        break;
+
+      case R_AARCH64_LDST128_ABS_LO12_NC:
+        {
+          overflow_check = false;
+          ret = reloc_insn_imm(RELOC_OP_ABS, addr, val, 4, 8,
+                               INSN_IMM_12);
+        }
+        break;
+
+      case R_AARCH64_TSTBR14:
+        {
+          ret = reloc_insn_imm(RELOC_OP_PREL, addr, val, 2, 14,
+                               INSN_IMM_14);
+        }
+        break;
+
+      case R_AARCH64_CONDBR19:
+        {
+          ret = reloc_insn_imm(RELOC_OP_PREL, addr, val, 2, 19,
+                               INSN_IMM_19);
+        }
+        break;
+
+      case R_AARCH64_JUMP26:
+      case R_AARCH64_CALL26:
+        {
+          ret = reloc_insn_imm(RELOC_OP_PREL, addr, val, 2, 26,
+                               INSN_IMM_26);
+        }
+        break;
+
+      default:
+        berr("ERROR: Unsupported relocation: %d\n",
+             (int)ELF64_R_TYPE(rel->r_info));
```

Comment:
Why not to use `PRIu64` here the same as at line 786?

Reply:
Done~